velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,231 @@
+      int32_t limit = std::numeric_limits<int32_t>::max();
+      if (limits) {
+        const auto constant = limits->valueAt<int32_t>(0);
+        if (constant > 0) {
+          limit = constant;
+        }
+      }
+      if (delim.empty()) {
+        rows.applyToSelected([&](vector_size_t row) {
+          splitEmptyDelimiter(rawStrings[row], limit, row, resultWriter);
+        });
+      } else {
+        rows.applyToSelected([&](vector_size_t row) {
+          split(rawStrings[row], delim, limit, row, resultWriter);
+        });
+      }
+    } else {
+      rows.applyToSelected([&](vector_size_t row) {
+        const auto delim = delims->valueAt<StringView>(row);
+        int32_t limit = std::numeric_limits<int32_t>::max();
+        if (limits) {
+          const auto limitValue = limits->valueAt<int32_t>(row);
+          if (limitValue > 0) {
+            limit = limitValue;
+          }
+        }
+        if (delim.empty()) {
+          splitEmptyDelimiter(
+              strings->valueAt<StringView>(row), limit, row, resultWriter);
+        } else {
+          split(
+              strings->valueAt<StringView>(row),
+              delim,
+              limit,
+              row,
+              resultWriter);
+        }
+      });
+    }
 
     resultWriter.finish();
-
-    // Reference the input StringBuffers since we did not deep copy above.
+    // Ensure that our result elements vector uses the same string buffer as
+    // the input vector of strings.
     result->as<ArrayVector>()
         ->elements()
         ->as<FlatVector<StringView>>()
-        ->acquireSharedStringBuffers(args[0].get());
+        ->acquireSharedStringBuffers(strings->base());
   }
 
  private:
-  const char pattern_;
-};
-
-/// This class will be updated in the future as we support more variants of
-/// split
-class Split final : public exec::VectorFunction {
- public:
-  Split() {}
+  // When pattern is empty, split each character out. Since Spark 3.4, when
+  // delimiter is empty, the result does not include an empty tail string, e.g.
+  // split('abc', '') outputs ["a", "b", "c"] instead of ["a", "b", "c", ""].
+  // The result does not include remaining string when limit is smaller than the
+  // string size, e.g. split('abc', '', 2) outputs ["a", "b"] instead of ["a",
+  // "bc"].
+  void splitEmptyDelimiter(
+      const StringView current,
+      int32_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    resultWriter.setOffset(row);
+    auto& arrayWriter = resultWriter.current();
+    if (current.size() == 0) {
+      arrayWriter.add_item().setNoCopy(StringView());
+      resultWriter.commit();
+      return;
+    }
+
+    const char* const begin = current.begin();
+    const char* const end = current.end();
+    const char* pos = begin;
+    while (pos < end && pos < limit + begin) {
+      arrayWriter.add_item().setNoCopy(StringView(pos, 1));
+      pos += 1;
```

Comment:
I wonder if this works for unicode characters which could occupy more than 1 byte. Could you confirm and add some tests? Perhaps use `!utf_cont` to only split for the first byte.

Reply:
This function not work for unicode (non utf-8) characters, what do you means `Perhaps use !utf_cont to only split for the first byt` ? @rui-mo

Reply:
We could use `utf_cont` to check if a byte is continuous byte in a character. If no, the byte should belong to a new character and we can split it out.